### ISSUE TITLE
VS/MSBuild per-project integration that "just works"

### ISSUE THREAD

**Issue by davidmatson** [microsoft/vcpkg#23366]
**Is your feature request related to a problem? Please describe.**
I want to be able use vcpkg as a submodule and enable vcpkg in a vcxproj in Visual Studio without having to add any per-user prerequisites or extra per-repo setup steps.

**Proposed solution**
Update the vcpkg targets to support bootstrapping.

**Describe alternatives you've considered**
Requiring users to bootstrap manually, but that makes our repo harder-to-use than before.
Using the generated NuGet package via "integrate", but that adds a dependency on NuGet, which we're trying to minimize, and is not recommended via the docs.

**Additional context**
Here's an example workaround:

vcpkg_workaround.targets:
```xml
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Target Name="VcpkgBootstrap" BeforeTargets="ClCompile"
            Condition="'$(VcpkgEnabled)' == 'true'"
            Inputs="$(_ZVcpkgRoot)bootstrap-vcpkg.bat"
            Outputs="$(_ZVcpkgExecutable)">
        <Message Text="Bootstrapping vcpkg" Importance="High" />
        <Exec Command="%22$(_ZVcpkgRoot)bootstrap-vcpkg.bat%22" StandardOutputImportance="High" />
    </Target>

    <Import Project="$(MSBuildThisFileDirectory)\vcpkg\scripts\buildsystems\msbuild\vcpkg.targets" />
</Project>
```

And then each project just something like:
```xml
    <Import Project="$..\vcpkg_workaround.targets" />
```

I imagine this question has come up before, but I couldn't find a current open issue tracking. Overall, my hope is that for folks who use VS without CMake, there can be a way to keep the repo self-contained and still "just work" when the project is opened/built for the first time.

**Comment by JackBoosY** [comment]
Not likely.
Since some settings such as `triplet` and `is_static` are common to the solution, we always set all configurations to the solution instead of the projects.

**Comment by davidmatson** [comment]
It says here the triplet can set in the project:
https://github.com/microsoft/vcpkg/blob/2777eada3f0608c337eced841c0023ab405e03d8/docs/users/integration.md
> To override the automatically chosen [triplet](https://github.com/microsoft/vcpkg/blob/2777eada3f0608c337eced841c0023ab405e03d8/docs/users/triplets.md), you can specify the MSBuild property VcpkgTriplet in your .vcxproj.
I assume the same is true for VcpkgUseStatic.

**Comment by JackBoosY** [comment]
@BillyONeal What do you think about ?

**Comment by JulZimmermann** [comment]
Hey. I kind of have the same problem. I want to integrate vcpkg into our MSBuild solution in such a way, that my colleagues don't have to perform any additional steps to build the solution.

The documented "Import .props and .targets" approach from https://vcpkg.io/en/docs/users/buildsystems/msbuild-integration.html seems like a good way in combination with a git submodule of the vcpkg repo. (This approach is even mentioned on the documentation site.)

Unfortunately, this doesn't work, because (unlike with the CMake integration) the bootstrap script isn't called automatically.

A solution-wide (not per project) integration would be fine for me, but it's not clear to me how to set this up.

**Comment by BillyONeal** [comment]
@ras0219 @ras0219-msft indicates that this was not an intentional omission.

**Comment by ras0219-msft** [comment]
Some comments on the posted blob that should be considered before merging into vcpkg:
- It should be conditional on manifest mode
- It should be `BeforeTargets="VcpkgInstallManifestDependencies"` (or that one should depend upon this one)
- We should possibly add a conditional to disable the auto-bootstrap
- We should ensure that it (bootstrap-vcpkg.bat) is safe to call in parallel

**Comment by davidmatson** [comment]
> We should ensure that it (bootstrap-vcpkg.bat) is safe to call in parallel

Yes. We've been using the above blob and found that in practice that it (bootstrap-vcpkg.bat) is not safe to call in parallel. I suspect some kind of lock file or other concurrency control mechanism would be needed.

**Comment by JulZimmermann** [comment]
> > We should ensure that it (bootstrap-vcpkg.bat) is safe to call in parallel
> 
> Yes. We've been using the above blob and found that in practice that it (bootstrap-vcpkg.bat) is not safe to call in parallel. I suspect some kind of lock file or other concurrency control mechanism would be needed.

The CMake integration (via toolchain file) does call bootstrap automatically in manifest mode. So maybe a locking mechanism is implemented there already, that could be reused? Or maybe it is missing there also?

**Comment by BillyONeal** [comment]
>So maybe a locking mechanism is implemented there already, that could be reused?

There is not. There is one CMake configure step for a whole build ever, there are many many MSBuild invocations (for each `.vcxproj`)

**Comment by JulZimmermann** [comment]
Hi, is there some progress on auto-bootstrap with MSBuild?

Thanks!

**Comment by BillyONeal** [comment]
> Hi, is there some progress on auto-bootstrap with MSBuild?

I don't believe we plan to implement this.

However we are planning on bundling vcpkg with Visual Studio so you won't need to bootstrap at all there, at least for manifest mode.

**Comment by mschofie** [comment]
> > So maybe a locking mechanism is implemented there already, that could be reused?
>
> There is not. 

MSBuild does have a locking mechanism to use - it guarantees that MSBuild tasks with a target and a given set of global properties aren't executed in parallel. So if bootstrap.cmd were invoked from such a context, it would be run only once. Ensuring that the global properties are unique is tricky until [`IBuildEngine6`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.framework.ibuildengine6?view=msbuild-17-netcore) was introduced, which adds a method to get the global properties. So you could:

1. Add an 'outer target' - that's marked as `BeforeTargets="VcpkgInstallManifestDependencies"`, with `Inputs="$(_ZVcpkgRoot)\scripts\vcpkg-tool-metadata.txt"` and `Outputs="$(_ZVcpkgExecutable)"`.
2. Within the 'outer target', get the GlobalProperties, then kick-off an MSBuild task with an 'inner target', removing all of the GlobalProperties, and setting `_ZVcpkgRoot` to both lock on that value, and propagate the path through.
3. In the inner-target, Exec `$(_ZVcpkgRoot)\bootstrap-vcpkg.bat`.

I've pushed a hand-waving implementation [here](https://github.com/mschofie/vcpkg/commit/b68a00fe44b9ee28249e1d57ed40466280b6f17b).

Is something like this something that VCPkg could consider?

**Comment by Cheney-W** [comment]
@mschofie Could you please provide a PR for this issue? Thank you!

**Comment by mschofie** [comment]
I think my branch needs some more work - there's definitely some 'due dilligence' needed. I'll try and take a look today and send something out for discussion.
